README.md
# Jquery Asynchronous Image Loader (JAIL)

JAIL is a jQuery plugin that lazy load images making your page load faster. 

[![Build Status](https://secure.travis-ci.org/sebarmeli/JAIL.png)](http://travis-ci.org/sebarmeli/JAIL)

Images are downloaded when they are visible or when they become visible inside the viewport (area you see in your browser). Images can be loaded after an event is triggered (such as `click`, `mouseover`, and `scroll`) or after a specified delay.  It's advisable to call jail() after the DOM has been constructed (document ready).

## Getting Started

First of all, this plugin requires you to make some HTML changes. The `data-src` attribute (HTML5 data attribute) should contain the location of the image, instead the `src` attribute should contain a placeholder such as a really tiny image (E.g. 1 px x 1 px). Also, I would suggest to add a `noscript` block so that in case the user doesn't have Javascript enabled, the images will be displayed (progressive enhancement).

```html
	<img class="lazy" src="/img/blank.gif" data-src="/img/image1.jpg"/>
	<noscript>
		<img src="/img/image1.jpg"/>
	<noscript>
```

In a basic scenario, you just need to import `jquery`, `jail.js` and call the jail() function on the images you want to lazy load.

```html
	<script src="/lib/jquery.js"></script>
	<script src="/src/jail.js"></script>
	<script>
		$(function(){
			$('img.lazy').jail();
		});
	</script>
```

The images in the viewport are loaded straight away after the DOM is ready. As soon other images become visible in the viewport (for instance after the user scrolls down), they are lazy loaded.

## Config
You can add additional configuration options when you initially call jail():

- `timeout`     : number of msec after that the images will be loaded - Default: `10`
- `effect`      : any jQuery effect that makes the images display (e.g. "fadeIn") - Default: `NULL`

  **NOTE:** If you are loading a large number of images, it is best  NOT to use this setting.

- `speed`       : string or number determining how long the animation will run  - Default: 400
- `triggerElement`    : selector that triggers the images to be loaded - Default: `NULL`
- `event`       : event : event that triggers the image to be loaded. You can choose among `load`, `click`, `mouseover`, 'scroll', etc. Default: `load`
- `callback`    : function that will be called after all the images are loaded - Default: ""
- `callbackAfterEachImage`    : function that will be called after each image is loaded - Default: ""
- `placeholder` : location of an image (such a loader) you want to display while waiting for the images to be loaded - Default: ""
- `offset`      : an offset of "500" would cause any images that are less than 500px below the bottom of the window or 500px above the top of the window to load. - Default: 0
- `loadHiddenImages` : boolean to load hidden images - Default: false (so hidden images are not loaded)

## How to invoke jail()

Here are some examples on how to invoke jail() in order to have a better understanding of how the plugin works

### Load images after clicking on anchor with id 'link'. The images will fade in with speed 500. Placeholder specified.

```javascript
	$(function(){
		$('img.lazy').jail({
			triggerElement:'a#link',
			event: 'click',
			effect: 'fadeIn',
			speed : 500,
			placeholder : 'img/loader.gif',
			callback : SA.setActive
		});
	});
```

### Initially load the visible images within `#my_container`. Then, as you scroll inside `#my_container`, images become visible

```javascript
	$(function(){
		$('img.lazy').jail({
			triggerElement:'#my_container',
			event: 'scroll'
		});
	});
```

### After 1 second the hidden images are loaded

```javascript
	$(function(){
		$('img.lazy').jail({
			timeout : 1000
		});
	});
```

### Load images after mouse-overing on the placeholder for the image

```javascript
	$(function(){
		$('img.lazy').jail({
			event: 'mouseover',
			placeholder : 'img/loader.gif'
		});
	});
```

### Load the images that are up to 300px over/below theviewport

```javascript
	$(function(){
		$('img.lazy').jail({
			offset : 300
		});
	});
```

### Alert saying "all the images are loaded" is called after all the images are loaded, alert saying "one image is loaded" after one image is loaded 

```javascript
	$(function(){
		$('img.lazy').jail({
			callback : (function(){alert("All the images are loaded");}),
			callbackAfterEachImage : function() {alert("one image is loaded");}
		});
	});
```

### Ignore hidden images to be loaded (images with or under a "display:none" or with hidden "visibility" or not visible inside a "overflow:hidden" element)

```javascript
	$(function(){
		$('img.lazy').jail({
			loadHiddenImages : true
		});
	});
```

## Demo

You can view a few demo examples usign JAIL [here](https://github.com/sebarmeli/JAIL/tree/master/demo)

## AMD support

Plugin supports AMD through define() method. If you use RequireJS (version > 2.0), you can require the plugin as follows:

```javascript
requirejs.config({
    baseUrl: 'lib',
    paths: {
        app: 'src'
    },
    shim: {
        'app/jail': ['jquery']
    }
});

require(["jquery", "app/jail"], function() {
    $(function(){
		$('img.lazy').jail();
	});
});
```

For more information, you view example 14 and example 15 from the [demo folder](https://github.com/sebarmeli/JAIL/tree/master/demo)

## Testing / Building the plugin

After getting node and npm, install grunt and grunt-jasmine-runner.

```npm install grunt```
```npm install grunt-jasmine-runner```

You can run Jasmine specs through phantomjs with :

```grunt jasmine```

If you don't have phantomjs, please download it from [here](http://phantomjs.org/)

You can run JSHint, Jasmine specs and minification tool simply launching: ```grunt```

## Licence

Copyright (c) 2011-2012 Sebastiano Armeli-Battana
Licensed under the MIT license.
(https://github.com/sebarmeli/JAIL/blob/master/MIT-LICENSE.txt)


CHANGELOG.txt
JAIL 1.0.0 (14/10/2011) 
----------------------- -
- Important fix inside _isVisibleInContainer() aroudn checking if the parent of the image exists
- Grunt.js support
- Switched from YUITest to Jasmine
- AMD support improvements (see example 14-15)
- Travic CI support

JAIL 0.9.9 (30/12/2011) 
-----------------------	-
- fixed issue with callback
- fixed issue with the use of '$' (AMD case)
- 'scroll' event managed in a better way

JAIL 0.9.8 (29/12/2011) 
-----------------------	-
- 'data-src' attributes substitutes 'data-href'
- 'load' will replace 'load+scroll' ('+scroll' won't exist anymore)
- 'selector' attribute renamed 'triggerElement'
- 'loadHiddenImages' attribute replaces 'ignoreHiddenImages'
- 'package.json' file added
- Added support for AMD spec
- Added 'post-receive hook' to the plugins.jquery.com site (not functional yet)
- Added checks on the 'overflow' CSS property
- Removed "asynchImageLoader" function
- Code restructured in meaningful public and private functions
- Removed GPL version
- Added example using 'require.js'
- Fixed issue #17, #20, #21, #22
- Better way of handling unbinding events

- JAIL 0.9.7 (12/10/2011) - Issue 16 fixed, "ignoreHiddenImages" parameter added and "container" parameter removed.

- JAIL 0.9.5 (3/08/2011) - Issues around images visible inside a container or inside an iframe been fixed. JS filename changed into jail.js.

- JAIL 0.9 (13/05/2011) - callback fixes + support for callbackAfterEachImage parameter

- JAIL 0.8 (03/04/2011) - jail() function, fixed critical issue on v0.7, resizing function, scrolling fixes

- JAIL 0.7 (19/02/2011) - Added "offset" configuration

- JAIL 0.6 (05/02/2011) - Fixed a critical issue on "effect" configuration

- JAIL 0.5 (27/01/2011) - 'Delay' configuration removed, critical fix on unbind method, refactoring to pass JSLint without warnings.

- JAIL 0.4 (21/01/2011) - Added ability to detect scroll on containing elements instead of just `window`. Plus, lots of refactoring. (Contributor : Derek Lindahl - dlindahl)

- JAIL 0.2 (20/01/2011) - HTML5 data attribute `data-href` used instead of `name` attribute

src/jail.js
/*
*  JAIL: jQuery Asynchronous Image Loader
*
* Copyright (c) 2011=12 Sebastiano Armeli-Battana (http://www.sebastianoarmelibattana.com)
*
* By Sebastiano Armeli-Battana (@sebarmeli)
* Licensed under the MIT license.
* https://github.com/sebarmeli/JAIL/blob/master/MIT-LICENSE.txt
*
* Tested with jQuery 1.3.2+ on FF 2+, Opera 10+, Safari 4+, Chrome 8+ on Win/Mac/Linux
* and IE 6/7/8 on Win.
*
* Contributor : Derek Lindahl - @dlindahl
*
* @link http://github.com/sebarmeli/JAIL
* @author Sebastiano Armeli-Battana
* @date 14/10/2012
* @version 1.0.0
*
*/
;(function ( name, definition ){

	var theModule = definition(jQuery),
		hasDefine = typeof define === 'function' && define.amd;

	if ( hasDefine ){  // AMD module

		define( name , ['jquery'], theModule );
		
	}  else { // assign 'jail' to global objects
		
		( this.jQuery || this.$ || this )[name] = theModule;

	}
}( 'jail', function ($) {

	var $window = $( window ),
		
		// Defaults parameters
		defaults = {
			timeout : 1,
			effect : false,
			speed : 400,
			triggerElement: null,
			offset : 0,
			event : 'load',
			callback : null,
			callbackAfterEachImage : null,
			placeholder : false,
			loadHiddenImages : false
		},
		
		// current stack of images
		currentStack = [],
		
		// true if 'callback' fn is called
		isCallbackDone = false;
		
	/* 
	* Public function defining 'jail'
	*
	* @module jail
	* @param elems : images to load - jQuery elements
	* @param opts : configurations object
	*/
	$.jail = function( elems, opts ) {

		var elements = elems || {},
			options = $.extend( {}, defaults, opts );
	
		// Initialize plugin
		$.jail.prototype.init( elements, options );
		
		// When the event is not specified the images will be loaded with a delay
		if(/^(load|scroll)/.test( options.event )) {
			// 'load' event
			$.jail.prototype.later.call( elements, options );
		} else {
			$.jail.prototype.onEvent.call( elements, options );
		}
	};
	
	/* 
	* Method in charge of initializing the plugin, storing
	* the 'element triggering the image to load' in a data attribute 
	* for each image and displaying the placeholder image (if existing)
	*
	* @method init
	* @param {Array} elems Images to load - jQuery elements
	* @param {Object} opts Configurations object
	*/
	$.jail.prototype.init = function( elements, options ) {
		
		// Store the selector triggering jail into 'triggerElem' data for the images selected 
		elements.data("triggerElem", ( options.triggerElement ) ? $( options.triggerElement ) : $window );
		
		// Use a placeholder in case it is specified
		if ( !!options.placeholder ) {
			elements.each(function(){
				$(this).attr( "src", options.placeholder );
			});
		}
	};
	
		
	/* 
	* Function called when 'event' is different from "load" or "scroll". Two scenarios:
	* a) Element triggering the images to be loaded (events available on the element: "click", "mouseover", "scroll")
	* b) Event on the image itself triggering the image to be loaded
	*
	* @param options : configurations object
	*/
	$.jail.prototype.onEvent = function( options ) {
		var images = this;
		
		if (!!options.triggerElement) {
			
			// Event on the 'triggerElement' obj
			_bindEvent( options, images );

		} else {
			
			// Event on the image itself
			images.bind( options.event, {options: options, images: images}, function(e) {
				var $img = $(this),
					options = e.data.options,
					images = e.data.images;
				
				currentStack = $.extend( {}, images );
				
				// Load the image
				_loadImage( options, $img );

				// Image has been loaded so there is no need to listen anymore
				$(e.currentTarget).unbind( e.type );
			});
		}
	};
	
	/* 
	* Method called when "event" is equals to "load" (default) or "scroll". The visible images will be 
	* loaded after a specified timeout (or after 1 ms). The scroll method will be bound to the window
	* to load  the images not visible onload.
	*
	* @param options : configurations object
	*/
	$.jail.prototype.later = function( options ) {
		var images = this;

		// After [timeout] has elapsed, load the visible images
		setTimeout(function() {
			
			currentStack = $.extend( {}, images );
			
			//Load the visible ones
			images.each(function(){
				_loadImageIfVisible( options, this, images );
			});
			
			// When images become available (scrolling or resizing), they will be loaded 
			options.event = "scroll";
			_bindEvent( options, images );
			
		}, options.timeout);
	};
		
	/* 
	* Bind _bufferedEventListener() to the event on window/triggerElement. The handler is bound to 
	* resizing the window as well
	*
	* @param options : configurations object
	* @param images : images in the current stack
	*/
	function _bindEvent ( options, images ) {
		var triggerElem = false;

		if (!!images) {
			triggerElem = images.data("triggerElem");
		}
		
		// Check if there are images to load
		if (!!triggerElem && typeof triggerElem.bind === "function") {
			triggerElem.bind( options.event, {options:options, images : images}, _bufferedEventListener );
			$window.bind( 'resize', {options:options, images : images}, _bufferedEventListener );
		}
	}

	/* 
	* Remove any elements that have been loaded from the jQuery stack.
	* This should speed up subsequent calls by not having to iterate over the loaded elements.
	*
	* @param stack : current images stack
	*/
	function _purgeStack ( stack ) {
		// number of images not loaded
		var i = 0;

		if (stack.length === 0) { return; }

		// Check on existence of 'data-src' attribute to verify if the image has been loaded
		while(true) {
			if(i ===  stack.length) {
				break;
			} else {
				if (stack[i].getAttribute('data-src')) {
					i++;
				} else {
					stack.splice( i, 1 );
				}
			}
		}
	}

	/* 
	* Event handler for the images to be loaded. Function called when 
	* there is a triggerElement or when there are images to be loaded after scrolling 
	* or resizing window/container 
	*
	* @param e : event
	*/
	function _bufferedEventListener (e) {
		var images = e.data.images,
			options = e.data.options;

		images.data('poller', setTimeout(function() {
			
			currentStack = $.extend( {}, images );
			_purgeStack(currentStack);
			
			// Load only the images left
			$(currentStack).each(function (){
				if (this === window) {
					return;
				}
				_loadImageIfVisible(options, this, currentStack);
			});
			
			//Unbind when there are no images
			if ( _isAllImagesLoaded (currentStack) ) {
				$(e.currentTarget).unbind( e.type );
				return;
			} 
			// When images are not in the viewport, let's load them when they become available
			else if (options.event !== "scroll"){
			
				// When images become available (scrolling or resizing), they will be loaded 
				var container = (/scroll/i.test(options.event)) ? images.data("triggerElem") : $window;
			
				options.event = "scroll";
				images.data("triggerElem", container);
				_bindEvent( options, $(currentStack) );
			}
		}, options.timeout));
	}
	
	/* 
	* Check if all the images are loaded
	*
	* @param images : images under analysis
	* @return boolean
	*/
	function _isAllImagesLoaded ( images ) {
		var bool = true;
		
		$(images).each(function(){
			if ( !!$(this).attr("data-src") ) {
				bool = false;
			}
		});
		return bool;
	}

	/* 
	* Load the image if visible in the viewport
	*
	* @param options : configurations object
	* @param image : image under analysis
	* @param images : list of images to load
	*/
	function _loadImageIfVisible ( options, image, images ) {
		var $img = $(image),
			container = (/scroll/i.test(options.event)) ? images.data("triggerElem") : $window,
			isVisible = true;

		// If don't you want to load hidden images (default beahviour)
		if ( !options.loadHiddenImages ) {
			isVisible = _isVisibleInContainer( $img, container, options ) && $img.is(":visible");
		}
		
		// Load the image if it is not hidden and visible in the screen
		if( isVisible && _isInTheScreen( container, $img, options.offset ) ) {
			_loadImage( options, $img );
		}
	}
	
	/* 
	* Function that returns true if the image is visible inside the "window" (or specified container element)
	*
	* @param $ct : container - jQuery obj
	* @param $img : image selected - jQuery obj
	* @param optionOffset : offset
	*/
	function _isInTheScreen ( $ct, $img, optionOffset ) {
		var is_ct_window  = $ct[0] === window,
			ct_offset  = (is_ct_window ? { top:0, left:0 } : $ct.offset()),
			ct_top     = ct_offset.top + ( is_ct_window ? $ct.scrollTop() : 0),
			ct_left    = ct_offset.left + ( is_ct_window ? $ct.scrollLeft() : 0),
			ct_right   = ct_left + $ct.width(),
			ct_bottom  = ct_top + $ct.height(),
			img_offset = $img.offset(),
			img_width = $img.width(),
			img_height = $img.height();
		
		return (ct_top - optionOffset) <= (img_offset.top + img_height) &&
			(ct_bottom + optionOffset) >= img_offset.top &&
				(ct_left - optionOffset)<= (img_offset.left + img_width) &&
					(ct_right + optionOffset) >= img_offset.left;
	}

	/* 
	* Main function --> Load the images copying the "data-href" attribute into the "src" attribute
	*
	* @param options : configurations object
	* @param $img : image selected - jQuery obj
	*/
	function _loadImage ( options, $img ) {
		
		// Use cache Image object to show images only when ready
		var cache = new Image();

		cache.onload = function() {
			$img.hide().attr("src", cache.src);
			
			$img.removeAttr('data-src');
			// Images loaded with some effect if existing
			if( options.effect) {

				if ( options.speed ) {
					$img[options.effect](options.speed);
				} else {
					$img[options.effect]();
				}
				$img.css("opacity", 1);
				$img.show();
			} else {
				$img.show();
			}
		
			_purgeStack(currentStack);
		
			// Callback after each image is loaded
			if ( !!options.callbackAfterEachImage ) {
				options.callbackAfterEachImage.call( this, $img, options );
			}
			
			if ( _isAllImagesLoaded (currentStack) && !!options.callback && !isCallbackDone ) {
				options.callback.call($.jail, options);
				isCallbackDone = true;
			}
		};
		cache.src = $img.attr("data-src");
	}
		
	/* 
	* Return if the image is visible inside a "container" / window. There are checks around
	* "visibility" CSS property and around "overflow" property of the "container"
	*
	* @param $img : image selected - jQuery obj
	* @param container : container object
	* @param options : configurations object
	*/
	function _isVisibleInContainer ( $img, container, options ){

		var parent = $img.parent(),
			isVisible = true;

		while ( parent.length && parent.get(0).nodeName.toUpperCase() !== "BODY" ) {
			// Consider the 'overflow' property
			if ( parent.css("overflow") === "hidden" ) {
				if (!_isInTheScreen(parent, $img, options.offset)) {
					isVisible = false;
					break;
				}
			} else if ( parent.css("overflow") === "scroll" ) {
				if (!_isInTheScreen(parent, $img, options.offset)) {
					isVisible = false;
					$(currentStack).data("triggerElem", parent);
					
					options.event = "scroll";
					_bindEvent(options, $(currentStack));
					break;
				}
			}
			
			if ( parent.css("visibility") === "hidden" || $img.css("visibility") === "hidden" ) {
				isVisible = false;
				break;
			}
			
			// If container is not the window, and the parent is the container, exit from the loop
			if ( container !== $window && parent === container ) {
				break;
			}
			
			parent = parent.parent();
		}
		
		return isVisible;
	}

	// Small wrapper
	$.fn.jail = function( options ) {

		new $.jail( this, options );
		
		// Empty current stack
		currentStack = [];
		
		return this;
	};
	
	return $.jail;
}));